### PR TITLE
Use new protobuf API instead of deprecated one

### PR DESCRIPTION
**Pull request by keryell**
This fixes compilation failures like
```
  error: ‘int google::protobuf::MessageLite::ByteSize() const’ is deprecated: Please use ByteSizeLong() instead [-Werror=deprecated-declarations]
```
This should fix https://github.com/Xilinx/XRT/issues/4131

### CODE REVIEW

**Comment by keryell** [comment]
Anyone to improve the CI so we can get the reason of error, like compilation output?

**Comment by keryell** [comment]
Curious. How did the ARM64 ever pass?

**Comment by keryell** [comment]
Note that some scripts run in verbose mode to have some feedback from the CI.
I do not know whether this is requested by normal users.
Perhaps for the curious people?
Otherwise we could add a verbose option to be used at least by the CI.

**Review by stsoe — CHANGES_REQUESTED** [review]
@keryell Please comment on the build.sh and xrtdeps.sh changes and why the removed code is not necessary.

File: build/build.sh
Comment (rozumx) [inline build/build.sh:1]: Is this something we wish to do?
```
   -v        Print shell input lines as they are read.
   -x        Print commands and their arguments as they are executed.

```

File: src/runtime_src/core/include/xcl_api_macros.h
Comment (rozumx) [inline src/runtime_src/core/include/xcl_api_macros.h:35]: Now that the return type is now known, shouldn't the variable type also be updated.  

unsigned is unfortunately not a good 'type', shouldn't it be instead unsigned long.  For example:
```suggestion
     unsigned long c_len = c_msg.ByteSizeLong(); \
```

**Review by rozumx — CHANGES_REQUESTED** [review]
Question regarding:
1. The use of "-vx" for the bash scripts.
2. Not using "unsigned" type.

File: src/runtime_src/core/include/xcl_api_macros.h
Comment (keryell) [reply src/runtime_src/core/include/xcl_api_macros.h:35]: The code is probably wrong if the message size does not fit in an `int` like it was anyway wrong before if the message size was not fitting in the deprecated `ByteSize()`.
I do not have the big picture of the code so I do not know if it can happen or not.
Probably the code should get rid of macros, use more `auto` and functions/lambda with type-safety guarantees. But this sounds outside the scope of this...

File: build/build.sh
Comment (keryell) [reply build/build.sh:1]: Just a `--verbose` doing an equivalent `set`?

Comment (stsoe) [reply build/build.sh:1]: I failed to notice -vx.  I don't think we want that in our build script.   You can use -verbose to see the actual compile commands, no need for more verbosity.

**Comment by keryell** [comment]
> @keryell Please comment on the build.sh and xrtdeps.sh changes and why the removed code is not necessary.

The code looks like some work around for limitation in a 3 year old Linux distribution.
Should we keep a comment in the code about something that should not have happened at the first place?
Or do you mean a comment in the log message?

**Comment by stsoe** [comment]
> > @keryell Please comment on the build.sh and xrtdeps.sh changes and why the removed code is not necessary.
> 
> The code looks like some work around for limitation in a 3 year old Linux distribution.
> Should we keep a comment in the code about something that should not have happened at the first place?
> Or do you mean a comment in the log message?

We still have to support > 3 year old Linux distributions.  Unless it can proven that the check is completely obsolete for all the OSes Xilinx has to support, then let's leave it.

**Comment by keryell** [comment]
But something is very curious: it looks like the CI on the main branch of this repository has a completely different behavior from the CI I have on my branch.
So the same master branch is compiling on ARM64 Ubuntu bionic on the CI here while it would not compile on the CI of https://github.com/keryell/XRT
This is why I had to fix the ARM64 compilation... :-(
Am I missing something obvious here?

**Comment by stsoe** [comment]
> But something is very curious: it looks like the CI on the main branch of this repository has a completely different behavior from the CI I have on my branch.
> So the same master branch is compiling on ARM64 Ubuntu bionic on the CI here while it would not compile on the CI of https://github.com/keryell/XRT
> This is why I had to fix the ARM64 compilation... :-(
> Am I missing something obvious here?
Why is https://github.com/keryell/XRT 3939 commits behind master?

**Comment by keryell** [comment]
> Why is [keryell/XRT](https://github.com/keryell/XRT) 3939 commits behind master?

Because I do not use it? :-) I just push my branches there...
So I do not think this is the problem.

Look for someone else too: https://travis-ci.org/github/Xilinx/XRT/jobs/762502204
What is going on?

**Comment by keryell** [comment]
> We still have to support > 3 year old Linux distributions. Unless it can proven that the check is completely obsolete for all the OSes Xilinx has to support, then let's leave it.

I agree.
I made this "fix" to have ARM64 working and just understanding from the CI that it was not working at the first place, so there was no real support breakage.
But the current CI seems completely fishy now and thus the last 2 commits of this PR are just insanity caused by current CI state I do not understand...

**Comment by stsoe** [comment]
@keryell Can you change this PR and remove the changes to build.sh?   Whether the changes to build.sh are warranted or not, separating the two changes is preferred.

**Comment by keryell** [comment]
I completely forgot about this up to I tried to install the latest XRT today and it failed...
Here is the minimal change to have it to compile on a modern system.

File: src/runtime_src/core/include/xcl_api_macros.h
Comment (keryell) [reply src/runtime_src/core/include/xcl_api_macros.h]: I just put `auto` to implement your suggestion.

**Comment by stsoe** [comment]
ok to test

**Review by stsoe — APPROVED** [review]

**Comment by stsoe** [comment]
@keryell There are build errors, please check link per slacked message

**Comment by keryell** [comment]
OK, there were still 2 function calls to the deprecated ProtoBuf API.
Let's try to build again...

**Comment by keryell** [comment]
But what is curious is that it was compiling fine on our Ubuntu 20.10 machines...
Be ready for Ubuntu 21.04 coming soon! :-)

**Comment by stsoe** [comment]
> Be ready for Ubuntu 21.04 coming soon! :-)
We will of course make sure we can build on 21.04, but I believe official support is  LTS only.

**Comment by keryell** [comment]
So, according to https://github.com/eddelbuettel/rprotobuf/issues/70 it looks like the problem is more complex as `protoc` was changed with the safer behavior starting with version 3.6.
https://github.com/protocolbuffers/protobuf/blob/master/CHANGES.txt
Thus, using the 2 versions according to the version of the ProtoBuf used.
Last try before forking this repository and leaving this to the Jurassix world...

**Review by chvamshi-xilinx — APPROVED** [review]
Looks good. @venkatp-xilinx / @hemantk-xilinx/ @vboggara-xilinx. Can you please make this change in C-models/genericpcie model side as well

**Comment by chvamshi-xilinx** [comment]
Adding "don't merge" tag as this change required in device side as well. Ideally things should work, but better to verify both sw_emu and hw_emu before merging this change.

**Comment by keryell** [comment]
If someone spends some time to understand this, it might be an opportunity to check also what is going on in https://github.com/Xilinx/XRT/issues/5081

**Comment by chvamshi-xilinx** [comment]
Removed "donot merge" tag As @vboggara-xilinx & @hemantk-xilinx verified both sw_emu and hw_emu

**Pull request merged**
